File: frontend/lib/api/zoom.ts
import { apiClient } from './client';

export interface ZoomAccountOut {
  id: string;
  accountName: string;
  accountId?: string;
  clientId?: string;
  isDefault: boolean;
  createdAt?: string;
}

export interface ZoomClassOut {
  id: string;
  title: string;
  batchId: string;
  batchName?: string;
  teacherId: string;
  teacherName?: string;
  zoomMeetingUrl?: string;
  zoomStartUrl?: string;
  scheduledDate: string;
  scheduledTime: string;
  duration: number;
  durationDisplay?: string;
  status: string;
  zoomAccountId: string;
  createdAt?: string;
}

export interface PaginatedZoomClasses {
  data: ZoomClassOut[];
  total: number;
  page: number;
  perPage: number;
  totalPages: number;
}

export async function listAccounts(): Promise<ZoomAccountOut[]> {
  return apiClient('/zoom/accounts');
}

export async function createAccount(data: Record<string, any>): Promise<ZoomAccountOut> {
  return apiClient('/zoom/accounts', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

export async function updateAccount(accountId: string, data: Record<string, any>): Promise<ZoomAccountOut> {
  return apiClient(`/zoom/accounts/${accountId}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

export async function deleteAccount(accountId: string): Promise<void> {
  return apiClient(`/zoom/accounts/${accountId}`, { method: 'DELETE' });
}

export async function setDefaultAccount(accountId: string): Promise<ZoomAccountOut> {
  return apiClient(`/zoom/accounts/${accountId}/set-default`, { method: 'PATCH' });
}

export async function listClasses(params?: {
  batch_id?: string;
  status?: string;
  teacher_id?: string;
  page?: number;
  per_page?: number;
}): Promise<PaginatedZoomClasses> {
  return apiClient('/zoom/classes', { params: params as Record<string, string | number | undefined> });
}

export async function createClass(data: {
  title: string;
  batch_id: string;
  teacher_id: string;
  zoom_account_id: string;
  scheduled_date: string;
  scheduled_time: string;
  duration: number;
}): Promise<ZoomClassOut> {
  return apiClient('/zoom/classes', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

export async function updateClass(classId: string, data: Record<string, any>): Promise<any> {
  return apiClient(`/zoom/classes/${classId}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

export async function deleteClass(classId: string): Promise<void> {
  return apiClient(`/zoom/classes/${classId}`, { method: 'DELETE' });
}

export async function getFreshStartUrl(classId: string): Promise<{ startUrl: string; joinUrl: string }> {
  return apiClient(`/zoom/classes/${classId}/start-url`, { method: 'POST' });
}

export interface AttendanceItem {
  id: string;
  zoomClassId: string;
  studentId: string;
  studentName?: string;
  attended: boolean;
  joinTime?: string;
  leaveTime?: string;
  durationMinutes?: number;
}

export async function getAttendance(classId: string): Promise<AttendanceItem[]> {
  return apiClient(`/zoom/classes/${classId}/attendance`);
}

export async function syncAttendance(classId: string): Promise<{ synced: number }> {
  return apiClient(`/zoom/classes/${classId}/sync-attendance`, { method: 'POST' });
}

export async function getRecordings(classId: string) {
  return apiClient(`/zoom/classes/${classId}/recordings`);
}

// --- Global Recordings ---

export interface RecordingItem {
  id: string;
  classTitle: string;
  title?: string;
  description?: string;
  teacherName?: string;
  batchName?: string;
  scheduledDate?: string;
  scheduledTime?: string;
  thumbnailUrl?: string;
  duration?: number;
  fileSize?: number;
  status: string;
  deletedAt?: string;
  createdAt?: string;
}

export interface PaginatedRecordings {
  data: RecordingItem[];
  total: number;
  page: number;
  perPage: number;
  totalPages: number;
}

export async function listRecordings(params?: {
  page?: number;
  per_page?: number;
  include_deleted?: boolean;
  batch_id?: string;
}): Promise<PaginatedRecordings> {
  return apiClient('/zoom/recordings', { params: params as Record<string, string | number | undefined> });
}

export async function getRecordingSignedUrl(recordingId: string): Promise<{ url: string; type: string; expiresAt?: number }> {
  return apiClient(`/zoom/recordings/${recordingId}/signed-url`, { method: 'POST' });
}

export async function updateRecording(id: string, data: { title?: string; description?: string }): Promise<RecordingItem> {
  return apiClient(`/zoom/recordings/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

export async function deleteRecording(id: string): Promise<void> {
  return apiClient(`/zoom/recordings/${id}`, { method: 'DELETE' });
}

export async function deleteRecordingPermanent(id: string): Promise<void> {
  return apiClient(`/zoom/recordings/${id}/permanent`, { method: 'DELETE' });
}

export async function restoreRecording(id: string): Promise<RecordingItem> {
  return apiClient(`/zoom/recordings/${id}/restore`, { method: 'POST' });
}

export interface ZoomAnalytics {
  totalClasses: number;
  upcomingClasses: number;
  liveClasses: number;
  completedClasses: number;
  totalRecordings: number;
  averageAttendanceRate: number;
}

export async function getZoomAnalytics(): Promise<ZoomAnalytics> {
  return apiClient('/zoom/analytics');
}

File: backend/app/schemas/zoom.py
import re
import uuid
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel, field_validator


class ZoomAccountCreate(BaseModel):
    account_name: str
    account_id: str
    client_id: str
    client_secret: str
    is_default: bool = False


class ZoomAccountUpdate(BaseModel):
    account_name: Optional[str] = None
    account_id: Optional[str] = None
    client_id: Optional[str] = None
    client_secret: Optional[str] = None
    is_default: Optional[bool] = None


class ZoomAccountOut(BaseModel):
    id: uuid.UUID
    account_name: str
    account_id: Optional[str] = None
    client_id: Optional[str] = None
    is_default: bool
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ZoomAccountAdminOut(ZoomAccountOut):
    client_secret_masked: Optional[str] = None


class ZoomClassCreate(BaseModel):
    title: str
    batch_id: uuid.UUID
    teacher_id: uuid.UUID
    zoom_account_id: uuid.UUID
    scheduled_date: date
    scheduled_time: str
    duration: int

    @field_validator('scheduled_time')
    @classmethod
    def validate_time_format(cls, v: str) -> str:
        if not re.match(r'^\d{2}:\d{2}$', v):
            raise ValueError('scheduled_time must be in HH:MM format')
        h, m = int(v[:2]), int(v[3:5])
        if h > 23 or m > 59:
            raise ValueError('Invalid time value')
        return v


class ZoomClassUpdate(BaseModel):
    title: Optional[str] = None
    scheduled_date: Optional[date] = None
    scheduled_time: Optional[str] = None
    duration: Optional[int] = None


class ZoomClassOut(BaseModel):
    id: uuid.UUID
    title: str
    batch_id: uuid.UUID
    batch_name: Optional[str] = None
    teacher_id: uuid.UUID
    teacher_name: Optional[str] = None
    zoom_meeting_url: Optional[str] = None
    zoom_start_url: Optional[str] = None
    scheduled_date: date
    scheduled_time: str
    duration: int
    duration_display: Optional[str] = None
    status: str
    zoom_account_id: uuid.UUID
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class AttendanceOut(BaseModel):
    id: uuid.UUID
    zoom_class_id: uuid.UUID
    student_id: uuid.UUID
    student_name: Optional[str] = None
    attended: bool
    join_time: Optional[datetime] = None
    leave_time: Optional[datetime] = None
    duration_minutes: Optional[int] = None

    model_config = {"from_attributes": True}


class RecordingOut(BaseModel):
    id: uuid.UUID
    zoom_class_id: uuid.UUID
    bunny_video_id: Optional[str] = None
    duration: Optional[int] = None
    file_size: Optional[int] = None
    status: str
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class RecordingUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None


class RecordingListOut(BaseModel):
    id: uuid.UUID
    class_title: str
    title: Optional[str] = None
    description: Optional[str] = None
    teacher_name: Optional[str] = None
    batch_name: Optional[str] = None
    scheduled_date: Optional[date] = None
    scheduled_time: Optional[str] = None
    thumbnail_url: Optional[str] = None
    duration: Optional[int] = None
    file_size: Optional[int] = None
    status: str
    deleted_at: Optional[datetime] = None
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class RecordingSignedUrlOut(BaseModel):
    url: str
    type: str
    expires_at: Optional[int] = None  # Unix timestamp when token expires
